File: SmartCard-API-SDK-trial/eGK/eGK/HealthCardForm.cs
// --------------------------------------------------------------------------------------------
// HelloMctForm.cs
// Subsembly SmartCard API
// Copyright © 2004-2017 CardWerk Technologies
// --------------------------------------------------------------------------------------------
// NOTES
// To access "old" KVK storage card (SLE4442 or similar) CT-API is required. Use CardTerminalConfigurator 
// to make vendor-proprietary CT-API DLL available to SmartCardAPI
// OMNIKEY: ctdeuin.dll is 32 bit windows DLL: requires both HelloMct and CardTerminalConfigurator built for x86 target
//
using System;
using System.Diagnostics;
using System.IO;
using System.Windows.Forms;
using System.Text;
using Subsembly.SmartCard;
using System.Reflection;


namespace GermanHealthInsuranceCard
{
	/// <summary>
	/// Sample application that reads the data from a German health card. Both, KVK storage card and eGK 
    /// processor cards are supported.
	/// </summary>

	public class HealthCardForm : System.Windows.Forms.Form
	{
		private System.Windows.Forms.LinkLabel m_aLinkLabel;
		private System.Windows.Forms.Label m_aCopyrightLabel;
		private System.Windows.Forms.Label m_aPromptLabel;
		private System.Windows.Forms.Label nameLabel;
		private System.ComponentModel.Container components = null;
		private System.Windows.Forms.TextBox nameTextBox;
		private System.Windows.Forms.TextBox insuranceTextBox;
		private System.Windows.Forms.Label insuranceLabel;
		private System.Windows.Forms.Label addressLabel;
		private System.Windows.Forms.TextBox addressTextBox;
		private System.Windows.Forms.Label birthdayLabel;
		private Label insuranceNumberLabel;
		private TextBox insuranceNumberTextBox;
		private TextBox numberTextBox;
		private Label numberLabel;
        private Label m_aApiVersion;
		private System.Windows.Forms.TextBox birthdayTextBox;

		/// <summary>
		/// Standard no brain constructor.
		/// </summary>

		public HealthCardForm()
		{
			InitializeComponent();
            this.Text = "eGK & KVK sample rev. 11OCT2020";
            m_aApiVersion.Text = SMARTCARDAPI.ApiVersionInfo;
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>

		protected override void Dispose( bool fDisposing )
		{
			if (fDisposing)
			{
				if (components != null)
				{
					components.Dispose();
				}
			}
			base.Dispose(fDisposing);
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
            System.ComponentModel.ComponentResourceManager resources = new System.ComponentModel.ComponentResourceManager(typeof(HealthCardForm));
            this.m_aLinkLabel = new System.Windows.Forms.LinkLabel();
            this.m_aCopyrightLabel = new System.Windows.Forms.Label();
            this.m_aPromptLabel = new System.Windows.Forms.Label();
            this.nameLabel = new System.Windows.Forms.Label();
            this.nameTextBox = new System.Windows.Forms.TextBox();
            this.insuranceTextBox = new System.Windows.Forms.TextBox();
            this.insuranceLabel = new System.Windows.Forms.Label();
            this.addressLabel = new System.Windows.Forms.Label();
            this.addressTextBox = new System.Windows.Forms.TextBox();
            this.birthdayLabel = new System.Windows.Forms.Label();
            this.birthdayTextBox = new System.Windows.Forms.TextBox();
            this.insuranceNumberLabel = new System.Windows.Forms.Label();
            this.insuranceNumberTextBox = new System.Windows.Forms.TextBox();
            this.numberTextBox = new System.Windows.Forms.TextBox();
            this.numberLabel = new System.Windows.Forms.Label();
            this.m_aApiVersion = new System.Windows.Forms.Label();
            this.SuspendLayout();
            // 
            // m_aLinkLabel
            // 
            this.m_aLinkLabel.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Bottom | System.Windows.Forms.AnchorStyles.Right)));
            this.m_aLinkLabel.AutoSize = true;
            this.m_aLinkLabel.Location = new System.Drawing.Point(336, 308);
            this.m_aLinkLabel.Name = "m_aLinkLabel";
            this.m_aLinkLabel.Size = new System.Drawing.Size(156, 13);
            this.m_aLinkLabel.TabIndex = 7;
            this.m_aLinkLabel.TabStop = true;
            this.m_aLinkLabel.Text = "http://www.smartcard-api.com/";
            this.m_aLinkLabel.LinkClicked += new System.Windows.Forms.LinkLabelLinkClickedEventHandler(this.LinkLabelLinkClicked);
            // 
            // m_aCopyrightLabel
            // 
            this.m_aCopyrightLabel.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Bottom | System.Windows.Forms.AnchorStyles.Left)));
            this.m_aCopyrightLabel.AutoSize = true;
            this.m_aCopyrightLabel.Location = new System.Drawing.Point(16, 308);
            this.m_aCopyrightLabel.Name = "m_aCopyrightLabel";
            this.m_aCopyrightLabel.Size = new System.Drawing.Size(223, 13);
            this.m_aCopyrightLabel.TabIndex = 6;
            this.m_aCopyrightLabel.Text = "Copyright 2004-2020 CardWerk Technologies";
            // 
            // m_aPromptLabel
            // 
            this.m_aPromptLabel.Font = new System.Drawing.Font("Trebuchet MS", 12F, ((System.Drawing.FontStyle)((System.Drawing.FontStyle.Bold | System.Drawing.FontStyle.Italic))), System.Drawing.GraphicsUnit.Point, ((byte)(0)));
            this.m_aPromptLabel.Location = new System.Drawing.Point(16, 16);
            this.m_aPromptLabel.Name = "m_aPromptLabel";
            this.m_aPromptLabel.Size = new System.Drawing.Size(488, 48);
            this.m_aPromptLabel.TabIndex = 8;
            this.m_aPromptLabel.Text = "Please insert KVK or eHealth card...";
            // 
            // nameLabel
            // 
            this.nameLabel.Location = new System.Drawing.Point(16, 132);
            this.nameLabel.Name = "nameLabel";
            this.nameLabel.Size = new System.Drawing.Size(152, 20);
            this.nameLabel.TabIndex = 9;
            this.nameLabel.Text = "Name of  insured:";
            this.nameLabel.TextAlign = System.Drawing.ContentAlignment.MiddleLeft;
            // 
            // nameTextBox
            // 
            this.nameTextBox.Location = new System.Drawing.Point(184, 133);
            this.nameTextBox.Name = "nameTextBox";
            this.nameTextBox.Size = new System.Drawing.Size(304, 20);
            this.nameTextBox.TabIndex = 10;
            // 
            // insuranceTextBox
            // 
            this.insuranceTextBox.Location = new System.Drawing.Point(184, 72);
            this.insuranceTextBox.Name = "insuranceTextBox";
            this.insuranceTextBox.Size = new System.Drawing.Size(304, 20);
            this.insuranceTextBox.TabIndex = 12;
            // 
            // insuranceLabel
            // 
            this.insuranceLabel.Location = new System.Drawing.Point(16, 72);
            this.insuranceLabel.Name = "insuranceLabel";
            this.insuranceLabel.Size = new System.Drawing.Size(152, 20);
            this.insuranceLabel.TabIndex = 11;
            this.insuranceLabel.Text = "Name of insurance:";
            this.insuranceLabel.TextAlign = System.Drawing.ContentAlignment.MiddleLeft;
            // 
            // addressLabel
            // 
            this.addressLabel.Location = new System.Drawing.Point(16, 193);
            this.addressLabel.Name = "addressLabel";
            this.addressLabel.Size = new System.Drawing.Size(152, 20);
            this.addressLabel.TabIndex = 14;
            this.addressLabel.Text = "Address of insured:";
            this.addressLabel.TextAlign = System.Drawing.ContentAlignment.MiddleLeft;
            // 
            // addressTextBox
            // 
            this.addressTextBox.Location = new System.Drawing.Point(184, 193);
            this.addressTextBox.Multiline = true;
            this.addressTextBox.Name = "addressTextBox";
            this.addressTextBox.Size = new System.Drawing.Size(304, 48);
            this.addressTextBox.TabIndex = 15;
            // 
            // birthdayLabel
            // 
            this.birthdayLabel.Location = new System.Drawing.Point(16, 254);
            this.birthdayLabel.Name = "birthdayLabel";
            this.birthdayLabel.Size = new System.Drawing.Size(152, 20);
            this.birthdayLabel.TabIndex = 16;
            this.birthdayLabel.Text = "Birthday of  insured:";
            this.birthdayLabel.TextAlign = System.Drawing.ContentAlignment.MiddleLeft;
            // 
            // birthdayTextBox
            // 
            this.birthdayTextBox.Location = new System.Drawing.Point(184, 255);
            this.birthdayTextBox.Name = "birthdayTextBox";
            this.birthdayTextBox.Size = new System.Drawing.Size(304, 20);
            this.birthdayTextBox.TabIndex = 17;
            // 
            // insuranceNumberLabel
            // 
            this.insuranceNumberLabel.AutoSize = true;
            this.insuranceNumberLabel.Location = new System.Drawing.Point(16, 102);
            this.insuranceNumberLabel.Name = "insuranceNumberLabel";
            this.insuranceNumberLabel.Size = new System.Drawing.Size(108, 13);
            this.insuranceNumberLabel.TabIndex = 18;
            this.insuranceNumberLabel.Text = "Number of insurance:";
            // 
            // insuranceNumberTextBox
            // 
            this.insuranceNumberTextBox.Location = new System.Drawing.Point(184, 99);
            this.insuranceNumberTextBox.Name = "insuranceNumberTextBox";
            this.insuranceNumberTextBox.Size = new System.Drawing.Size(304, 20);
            this.insuranceNumberTextBox.TabIndex = 19;
            // 
            // numberTextBox
            // 
            this.numberTextBox.Location = new System.Drawing.Point(184, 159);
            this.numberTextBox.Name = "numberTextBox";
            this.numberTextBox.Size = new System.Drawing.Size(304, 20);
            this.numberTextBox.TabIndex = 20;
            // 
            // numberLabel
            // 
            this.numberLabel.AutoSize = true;
            this.numberLabel.Location = new System.Drawing.Point(16, 162);
            this.numberLabel.Name = "numberLabel";
            this.numberLabel.Size = new System.Drawing.Size(96, 13);
            this.numberLabel.TabIndex = 21;
            this.numberLabel.Text = "Number of insured:";
            // 
            // m_aApiVersion
            // 
            this.m_aApiVersion.AutoSize = true;
            this.m_aApiVersion.Location = new System.Drawing.Point(17, 286);
            this.m_aApiVersion.Name = "m_aApiVersion";
            this.m_aApiVersion.Size = new System.Drawing.Size(41, 13);
            this.m_aApiVersion.TabIndex = 22;
            this.m_aApiVersion.Text = "version";
            // 
            // HealthCardForm
            // 
            this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
            this.ClientSize = new System.Drawing.Size(512, 337);
            this.Controls.Add(this.m_aApiVersion);
            this.Controls.Add(this.numberLabel);
            this.Controls.Add(this.numberTextBox);
            this.Controls.Add(this.insuranceNumberTextBox);
            this.Controls.Add(this.insuranceNumberLabel);
            this.Controls.Add(this.birthdayTextBox);
            this.Controls.Add(this.addressTextBox);
            this.Controls.Add(this.insuranceTextBox);
            this.Controls.Add(this.nameTextBox);
            this.Controls.Add(this.m_aLinkLabel);
            this.Controls.Add(this.m_aCopyrightLabel);
            this.Controls.Add(this.birthdayLabel);
            this.Controls.Add(this.addressLabel);
            this.Controls.Add(this.insuranceLabel);
            this.Controls.Add(this.nameLabel);
            this.Controls.Add(this.m_aPromptLabel);
            this.Icon = ((System.Drawing.Icon)(resources.GetObject("$this.Icon")));
            this.Name = "HealthCardForm";
            this.Text = "eGK & KVK sample code";
            this.ResumeLayout(false);
            this.PerformLayout();

		}
		#endregion

		/// <summary>
		/// Entry point of application.
		/// </summary>

		[STAThread]
		static void Main() 
		{
			// Create a new text writer using the output stream, and add it to the trace listeners.
            string sPath = Environment.GetFolderPath(Environment.SpecialFolder.Personal);
            
            Stream aTraceFile = File.Create(Path.Combine(sPath, "HelloMctTrace.txt"));
			TextWriterTraceListener aListener = new TextWriterTraceListener(aTraceFile);
			Trace.Listeners.Add(aListener);

			try
			{
				HealthCardForm aHelloMctForm = new HealthCardForm();

				// Run the primary application form.

				Application.Run(aHelloMctForm);
			}
			catch (Exception x)
			{
				Trace.WriteLine(x.ToString());
			}
			finally
			{
				// Flush and close the trace output.

				Trace.Flush(); 
				aTraceFile.Flush();
				aTraceFile.Close();
			}
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="e"></param>

		protected override void OnLoad(EventArgs e)
		{
			// We attach card terminal event handlers before starting up the card terminal
			// manager. This ensures that we get a card inserted event for those cards that
			// are already inserted when this program is started.

			CardTerminalManager.Singleton.CardInsertedEvent +=
				new CardTerminalEventHandler(InsertedEvent);
			CardTerminalManager.Singleton.CardRemovedEvent +=
				new CardTerminalEventHandler(RemovedEvent);
            CardTerminalManager.Singleton.CardTerminalFoundEvent +=
                new CardTerminalEventHandler(TerminalFoundEvent);
            CardTerminalManager.Singleton.CardTerminalLostEvent +=
               new CardTerminalEventHandler(TerminalLostEvent);

			// Try to start up the card terminal manager. If this fails, then the application
			// is immediately aborted.

			if (!StartupCardTerminalManager())
			{
				Application.Exit();
			}

			base.OnLoad(e);
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="e"></param>

		protected override void OnClosed(EventArgs e)
		{
			// Whenever we did successfully call the CardTerminalManager Startup method,
			// we also must call the CardTerminalManager Shutdown method!

			if (CardTerminalManager.Singleton.StartedUp)
			{
				CardTerminalManager.Singleton.Shutdown();
			}

			base.OnClosed(e);
		}

		/// <summary>
		/// Helper that starts up the card terminal manager and cares about the case when no
		/// card terminals are installed.
		/// </summary>

		bool StartupCardTerminalManager()
		{
			bool fStartedUp = false;

			try
			{
				// Startup the SmartCard Subsembly. The parameter "true" means that any
				// PC/SC smart card readers will automatically be added to the smart card
				// configuration registry. If startup fails, then this will throw an
				// exception.

				int nCountReaders = CardTerminalManager.Singleton.Startup(true);
                fStartedUp = true;

			}
			catch (Exception x)
			{
				// TODO: Better diagnstic and error handling would be appropriate here.

				Trace.WriteLine(x.ToString());

				MessageBox.Show(
					"The card terminal manager could not be started up. This program will " +
					"be aborted now.",
					"HelloMCT",
					MessageBoxButtons.OK,
					MessageBoxIcon.Stop);
			}
			
			return fStartedUp;
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="aSender"></param>
		/// <param name="aEventArgs"></param>

		public void InsertedEvent(object aSender, CardTerminalEventArgs aEventArgs)
		{
			if (base.InvokeRequired)
			{
				object[] vParms = new object[2];
				vParms[0] = aSender;
				vParms[1] = aEventArgs;
				base.BeginInvoke(new CardTerminalEventHandler(InsertedEvent),
					vParms);
			}
			else
			{
				// Read and display card data of the inserted card. If multiple readers are
				// connected, and multiple cards are inserted, then this will overwrite the
				// display of the last insertion.

				this.ReadHealthCard(aEventArgs.Slot);
			}
		}
	
		/// <summary>
		/// 
		/// </summary>
		/// <param name="aSender"></param>
		/// <param name="aEventArgs"></param>
	
		public void RemovedEvent(object aSender, CardTerminalEventArgs aEventArgs)
		{
			if (base.InvokeRequired)
			{
				object[] vParms = new object[2];
				vParms[0] = aSender;
				vParms[1] = aEventArgs;
				base.BeginInvoke(new CardTerminalEventHandler(RemovedEvent),
					vParms);
			}
			else
			{
				// Display prompt and clear displayed card data. If multiple card readers
				// are connected and multiple cards are inserted, then the first card that
				// is removed will clear the display.

				this.PromptHealthCard();
			}
		}

        /// <summary>
        /// We add the reader to be monitored.
        /// </summary>
        /// <param name="aSender"></param>
        /// <param name="aEventArgs"></param>

        public void TerminalFoundEvent(object aSender, CardTerminalEventArgs aEventArgs)
        {
            Debug.WriteLine(aSender);
            if (base.InvokeRequired)
            {
                object[] vParms = new object[2];
                vParms[0] = aSender;
                vParms[1] = aEventArgs;
                base.BeginInvoke(new CardTerminalEventHandler(TerminalFoundEvent), vParms);
            }
            else
            {
                if (CardTerminalManager.Singleton.StartedUp)
                {
                    Debug.WriteLine("Found reader: " + aEventArgs.Slot.CardTerminalName);
                    this.m_aPromptLabel.Text = "Insert card ...";
                }
            }

        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="aSender"></param>
        /// <param name="aEventArgs"></param>

        public void TerminalLostEvent(object aSender, CardTerminalEventArgs aEventArgs)
        {
            Debug.WriteLine(aSender);
            if (base.InvokeRequired)
            {
                object[] vParms = new object[2];
                vParms[0] = aSender;
                vParms[1] = aEventArgs;
                base.BeginInvoke(new CardTerminalEventHandler(TerminalLostEvent),
                    vParms);
            }
            else
            {
                Debug.WriteLine("lost terminal " + aEventArgs.Slot.CardTerminalName);
                 // update number of readers
                 CardTerminalManager.Singleton.DelistCardTerminal(aEventArgs.Slot.CardTerminal); // remove from monitored list of readers
                 if (CardTerminalManager.Singleton.SlotCount == 0)
                 {
                     this.m_aPromptLabel.Text = "Connect reader ...";
                     // start looking for reader insertion
                     // done automatically by the singleton. The singleton raises an event if it 
                     // finds a new reader.
                 }
            }
        }



		/// <summary>
		/// 
		/// </summary>
		/// <param name="aSlot"></param>

		public void ReadHealthCard(CardTerminalSlot aSlot)
		{
			m_aPromptLabel.Text = "Reading card ...";
			m_aPromptLabel.Update();

			CardActivationResult nActivationResult;

			try
			{
				HealthCard aHealthCard;

                // Try the new eGK processor card.
                aHealthCard = _ReadEgkHealthCard(aSlot, out nActivationResult);

                // If not successful, try old memory card.
                if (aHealthCard == null)
                {

                    Debug.WriteLine("eGK power-up failed. Trying KVK storage card.");
                    aHealthCard = _ReadMctHealthCard(aSlot, out nActivationResult);
                    if (aHealthCard == null) Debug.WriteLine("KVK power-up also failed.");
                }

				if (aHealthCard != null)
				{
					m_aPromptLabel.Text = aHealthCard.CardName;
					m_aPromptLabel.Update();

					insuranceTextBox.Text = aHealthCard.InsuranceName;
					insuranceNumberTextBox.Text = aHealthCard.InsuranceNumber;

					StringBuilder sbName = new StringBuilder();
					if (aHealthCard.TitleName != null)
					{
						sbName.Append(aHealthCard.TitleName);
						sbName.Append(' ');
					}
					sbName.Append(aHealthCard.FirstName);
					sbName.Append(' ');
					sbName.Append(aHealthCard.LastName);
					nameTextBox.Text = sbName.ToString();

					numberTextBox.Text = aHealthCard.InsuredNumber;

					StringBuilder sbAddress = new StringBuilder();
					if (aHealthCard.Street != null)
					{
						sbAddress.AppendLine(aHealthCard.Street);
					}
					sbAddress.Append(aHealthCard.ZipCode);
					sbAddress.Append(' ');
					sbAddress.Append(aHealthCard.City);
					addressTextBox.Text = sbAddress.ToString();

					birthdayTextBox.Text = aHealthCard.Birthday.ToLongDateString();
				}
				else
				{
					Trace.WriteLine("CardActivationResult: " + nActivationResult.ToString());

					switch (nActivationResult)
					{
					case CardActivationResult.NoCard:
						this.PromptHealthCard();
						break;
					case CardActivationResult.UnresponsiveCard:
						m_aPromptLabel.Text = "Unresponsive Card.";
						break;
					case CardActivationResult.InUse:
						m_aPromptLabel.Text = "Card reader blocked by another application.";
						break;
					default:
						m_aPromptLabel.Text = "Inserted card is not supported!";
						break;
					}
				}
			}
			catch (CardTerminalException x)
			{
				switch (x.Code)
				{
				case CardTerminalExceptionCode.CardWithdrawn:
					this.PromptHealthCard();
					break;
				default:
					m_aPromptLabel.Text = "Card Terminal Exception: " + x.Code;
					break;
				}
			}
		}

		/// <summary>
		/// 
		/// </summary>

		void PromptHealthCard()
		{
			m_aPromptLabel.Text = "Please insert a German health insurance card...";
			insuranceTextBox.Text = null;
			insuranceNumberTextBox.Text = null;
			nameTextBox.Text = null;
			numberTextBox.Text = null;
			addressTextBox.Text = null;
			birthdayTextBox.Text = null;
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="aSlot"></param>
		/// <param name="nActivationResult"></param>
		/// <returns></returns>

		HealthCard _ReadMctHealthCard(
			CardTerminalSlot aSlot,
			out CardActivationResult nActivationResult)
		{
			HealthCard_KVK aMctHealthCard = new HealthCard_KVK();
			nActivationResult = aMctHealthCard.Read(aSlot);

			if (nActivationResult == CardActivationResult.Success)
			{
				return aMctHealthCard;
			}
            else if (nActivationResult == CardActivationResult.UnsupportedCard)
            {

                m_aPromptLabel.Text = "Unsupported card - Please insert a German health insurance card..."; 
                return null;
            }
			else
			{
				return null;
			}
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="aSlot"></param>
		/// <param name="nActivationResult"></param>
		/// <returns></returns>

		HealthCard _ReadEgkHealthCard(
			CardTerminalSlot aSlot,
			out CardActivationResult nActivationResult)
		{
			HealthCard_eGK aEgkHealthCard = new HealthCard_eGK();
			nActivationResult = aEgkHealthCard.Read(aSlot);

			if (nActivationResult == CardActivationResult.Success)
			{
				return aEgkHealthCard;
			}
			else
			{
				return null;
			}
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="sender"></param>
		/// <param name="e"></param>
	
		void LinkLabelLinkClicked(object sender, System.Windows.Forms.LinkLabelLinkClickedEventArgs e)
		{
			Process.Start(m_aLinkLabel.Text);
		}
	}
}

File: SmartCard-API-SDK-trial/eGK/eGK.VB/HealthCardForm.vb
' CardWerk SmartCard API
' Copyright © 2004-2013 CardWerk Technologies
' HIST
'27NOV2013 MJ add support for protected transaction


Option Strict On
Option Explicit On

Imports System.IO
Imports Subsembly.SmartCard
Imports System.Text


''' <summary>
''' The goal of the example to show how to connect to smart card using SmartCard API. 
''' </summary>
''' <remarks></remarks>
Public Class HealthCardForm


#Region "Class Members"
    Private WithEvents CardManager As CardTerminalManager
    Private Shared ReadOnly FILE_NAME As String = "HelloCardTrace.txt"
    Private aTraceFile As Stream
#End Region

#Region "Initialization"

    Public Sub New()

        ' This call is required by the Windows Form Designer.
        InitializeComponent()

        ' Add any initialization after the InitializeComponent() call.
        RegisterTraceListener()

        ' initialize card terminal manager
        InitializeCardManager()

    End Sub


    ''' <summary>
    ''' Register trace listener with path User_Folder/Documents/HelloCardTrace.txt
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub RegisterTraceListener()

        ' Create a new text writer using the output stream, and add it to the trace
        ' listeners.
        Dim sPath As String = Environment.GetFolderPath(Environment.SpecialFolder.Personal)
        aTraceFile = File.Create(Path.Combine(sPath, FILE_NAME))
        Dim aListener As TextWriterTraceListener = New TextWriterTraceListener(aTraceFile)
        Trace.Listeners.Add(aListener)


    End Sub

    Private Sub InitializeCardManager()

        Trace.WriteLine("Initialiazing card manager")

        ' QUESTION Why cardterminal manager is a singleton
        ' ANSWER -There is always only a single instance of the CardTerminalManager class, 
        ' hence it is called a singleton. 
        ' Applications must use the public Singleton property in order to get access 
        ' to the card terminal manager.

        CardManager = CardTerminalManager.Singleton

    End Sub

#End Region


#Region "Card Manager events handlers"


    ''' <summary>
    ''' Event handler when card is inserted
    ''' </summary>
    ''' <param name="aSender">CardTerminalManager</param>
    ''' <param name="aEventArgs">This class provides the event information for the card terminal events 
    ''' that are raised by the CardTerminalManager whenever a smart card is inserted or is removed. 
    ''' </param>
    Protected Sub InsertedEvent(ByVal aSender As Object, ByVal aEventArgs As CardTerminalEventArgs) _
        Handles CardManager.CardInsertedEvent

        If (InvokeRequired = True) Then

            Dim vParms(1) As Object
            vParms(0) = aSender
            vParms(1) = aEventArgs
            BeginInvoke(New CardTerminalEventHandler(AddressOf InsertedEvent), vParms)

        Else

            ' We catch any exceptions during card I/O. This is particularly important
            ' for fuzzy communication conditions. Example: an contactless card that 
            ' is not in the field throuout the whole I/O might cause an error on the 
            ' within the unmanaged Windows code. SmartCardAPI catches this in a general 
            ' exception
            Try
                Trace.WriteLine("Card inserted in reader.")
                ReadHealthCard(aEventArgs.Slot)

            Catch ex As Exception
                Trace.WriteLine(String.Format("Error reading the card: {0}", ex.Message))
                PromptHealthCard()
            End Try

        End If

    End Sub

    ''' <summary>
    ''' Event handler when card is removed
    ''' </summary>
    ''' <param name="aSender">CardTerminalManager</param>
    ''' <param name="aEventArgs">This class provides the event information for the card terminal events 
    ''' that are raised by the CardTerminalManager whenever a smart card is inserted or is removed. 
    ''' </param>
    Protected Sub RemovedEvent(ByVal aSender As Object, ByVal aEventArgs As CardTerminalEventArgs) _
        Handles CardManager.CardRemovedEvent

        If (InvokeRequired) Then

            Dim vParms(1) As Object
            vParms(0) = aSender
            vParms(1) = aEventArgs

            BeginInvoke(New CardTerminalEventHandler(AddressOf RemovedEvent), vParms)

        Else

            Trace.WriteLine("Card ejected from the card reader")
            PromptHealthCard()
        End If
    End Sub

#End Region




#Region "Windows From Events"

    Private Overloads Sub OnLoad(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        ' start the card terminal manager
        ' if start is failed or no card terminal found  - exit from app
        If (StartupCardTerminalManager() <> True) Then

            Application.Exit()

        End If

    End Sub

    Private Sub OnClose(ByVal sender As System.Object, ByVal e As System.Windows.Forms.FormClosedEventArgs) Handles MyBase.FormClosed

        ' shutdown card terminal manager
        ShutdownCardTerminalManager()

        ' close the trace file
        Trace.Flush()
        aTraceFile.Flush()
        aTraceFile.Close()

    End Sub

#End Region





#Region "Private functions"

    ' <summary>
    ' Helper that starts up the card terminal manager and cares about the case when no
    ' card terminals are installed.
    ' </summary>
    ''' <returns>Returns true if card terminal manager started sucessfulyy and card terminal 
    ''' slot is found, otherwise false</returns>
    Private Function StartupCardTerminalManager() As Boolean

        Dim fStartedUp As Boolean = False

        Trace.WriteLine("Starting card manager")

        Try
            ' Startup the SmartCard API. The parameter "true" means that any
            ' PC/SC smart card reader will automatically be added to the smart card
            ' configuration registry. If startup fails, then this will throw an
            ' exception.

            'QUESTION  what true means?
            Dim nCountReaders As Integer = CardManager.Startup(True)


            ' At least one card terminal is configured, enabled and was started up
            ' successfully. This is all we need, thus we can exit and return
            ' successfully.
            If (nCountReaders > 0) Then

                Trace.WriteLine(String.Format("{0} card terminals found.", nCountReaders))

                fStartedUp = True

            Else

                Trace.WriteLine(String.Format("{0} card terminals found.", nCountReaders))

                ' If the returned count of installed and enabled readers is zero, then it
                ' does not make sense to continue. Anyway we are obliged to call the
                ' CardTerminalManager Shutdown method.
                ShutdownCardTerminalManager()

                MessageBox.Show("No card reader/terminal available on this system. Please verify" & vbNewLine & _
                  "your smart card system. Check CardTerminal Configurator if available. ", _
                  "HelloCard", MessageBoxButtons.OK, MessageBoxIcon.Warning)

            End If

        Catch ex As Exception

            Trace.WriteLine(String.Format("StartCardTerminalManager Error: {0}", ex.Message))

            MessageBox.Show("Unable to run CardTerminalConfigurator. Will " & vbNewLine & _
              "exit this application.", "HelloCard", MessageBoxButtons.OK, _
              MessageBoxIcon.Stop, MessageBoxDefaultButton.Button1)

            fStartedUp = False
        End Try

        Return fStartedUp

    End Function


    ''' <summary>
    ''' Shutdown Card Terminal Manager
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub ShutdownCardTerminalManager()

        ' Every successful CardTerminalManager Startup method call,
        ' requires a mandatory CardTerminalManager Shutdown method call!
        If (CardManager.StartedUp = True) Then

            Trace.WriteLine("Shutdown card manager")
            CardManager.Shutdown()

        End If
    End Sub



    ''' <summary>
    ''' 
    ''' </summary>
    ''' <param name="aSlot"></param>

    Public Sub ReadHealthCard(ByVal aSlot As CardTerminalSlot)
        m_aPromptLabel.Text = "Reading card ..."
        m_aPromptLabel.Update()

        Dim nActivationResult As CardActivationResult

        Try
            Dim aHealthCard As SampleCode_eGK.HealthCard
            aHealthCard = _ReadEgkHealthCard(aSlot, nActivationResult)
            ' If not successful, try the new eGK processor card.

            If aHealthCard Is Nothing Then
                Debug.WriteLine("eGK power-up failed. Trying KVK storage card.")
                aHealthCard = _ReadMctHealthCard(aSlot, nActivationResult)
                If aHealthCard Is Nothing Then
                    Debug.WriteLine("KVK power-up also failed.")
                End If

            End If


            If aHealthCard IsNot Nothing Then
                m_aPromptLabel.Text = aHealthCard.CardName
                m_aPromptLabel.Update()

                insuranceTextBox.Text = aHealthCard.InsuranceName
                insuranceNumberTextBox.Text = aHealthCard.InsuranceNumber

                Dim sbName As New StringBuilder()
                If aHealthCard.TitleName IsNot Nothing Then
                    sbName.Append(aHealthCard.TitleName)
                    sbName.Append(" "c)
                End If
                sbName.Append(aHealthCard.FirstName)
                sbName.Append(" "c)
                sbName.Append(aHealthCard.LastName)
                nameTextBox.Text = sbName.ToString()

                numberTextBox.Text = aHealthCard.InsuredNumber

                Dim sbAddress As New StringBuilder()
                If aHealthCard.Street IsNot Nothing Then
                    sbAddress.AppendLine(aHealthCard.Street)
                End If
                sbAddress.Append(aHealthCard.ZipCode)
                sbAddress.Append(" "c)
                sbAddress.Append(aHealthCard.City)
                addressTextBox.Text = sbAddress.ToString()

                birthdayTextBox.Text = aHealthCard.Birthday.ToLongDateString()
            Else
                Trace.WriteLine("CardActivationResult: " + nActivationResult.ToString())

                Select Case nActivationResult
                    Case CardActivationResult.NoCard
                        Me.PromptHealthCard()
                        Exit Select
                    Case CardActivationResult.UnresponsiveCard
                        m_aPromptLabel.Text = "Unresponsive Card."
                        Exit Select
                    Case CardActivationResult.InUse
                        m_aPromptLabel.Text = "Card reader blocked by another application."
                        Exit Select
                    Case Else
                        m_aPromptLabel.Text = "Inserted card is not supported!"
                        Exit Select
                End Select
            End If
        Catch x As CardTerminalException
            Select Case x.Code
                Case CardTerminalExceptionCode.CardWithdrawn
                    Me.PromptHealthCard()
                    Exit Select
                Case Else
                    m_aPromptLabel.Text = "Card Terminal Exception: " + x.Message
                    Exit Select
            End Select
        End Try
    End Sub



    ''' <summary>
    ''' 
    ''' </summary>
    ''' <param name="aSlot"></param>
    ''' <param name="nActivationResult"></param>
    ''' <returns></returns>

    Private Function _ReadMctHealthCard(ByVal aSlot As CardTerminalSlot, ByVal nActivationResult As CardActivationResult) As SampleCode_eGK.HealthCard
        Dim aMctHealthCard As New SampleCode_eGK.MctHealthCard()
        nActivationResult = aMctHealthCard.Read(aSlot)

        If nActivationResult = CardActivationResult.Success Then
            Return aMctHealthCard
        Else
            Return Nothing
        End If
    End Function

    ''' <summary>
    ''' 
    ''' </summary>
    ''' <param name="aSlot"></param>
    ''' <param name="nActivationResult"></param>
    ''' <returns></returns>

    Private Function _ReadEgkHealthCard(ByVal aSlot As CardTerminalSlot, ByVal nActivationResult As CardActivationResult) As SampleCode_eGK.HealthCard
        Dim aEgkHealthCard As New SampleCode_eGK.EgkHealthCard()
        nActivationResult = aEgkHealthCard.Read(aSlot)

        If nActivationResult = CardActivationResult.Success Then
            Return aEgkHealthCard
        Else
            Return Nothing
        End If
    End Function


    ''' <summary>
    ''' 
    ''' </summary>

    Private Sub PromptHealthCard()
        m_aPromptLabel.Text = "Please insert card..."
        insuranceTextBox.Text = Nothing
        insuranceNumberTextBox.Text = Nothing
        nameTextBox.Text = Nothing
        numberTextBox.Text = Nothing
        addressTextBox.Text = Nothing
        birthdayTextBox.Text = Nothing
    End Sub



#End Region



End Class
